### ISSUE TITLE
From range projections to projected ranges

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8256]
From range projections to projected ranges<br>
<br>
https://ift.tt/3FDRBLC<br>
<br>
Oleksandr Koval<br>
<br>
<div>
<div>
<div>
<div>
<div>
<h2>Table of contents</h2>
<hr>
<h2>Introduction</h2>
<p>When I first watched range-related talks, I liked the idea of projections. I played with them a bit and still liked them. However, after trying to write range-based algorithms I found them not good enough and not pleasant to work with. In this post I&rsquo;ll explain why I don&rsquo;t like range projections in their current form and how I propose to fix them (demo implementation is provided).</p>
<p><em>Update. After this article was published, I received some feedback and realized that the proposed design has one problem. I described it in the section <a href="https://oleksandrkvl.github.io/2021/10/11/projected-ranges.html#the-flaw">Major flaw</a> and left the rest of the article untouched. Please keep that in mind while reading it. Thanks to all the people who shared their feedback and thoughts.</em></p>
<hr>
<h2>What a projection is</h2>
<p>If you are not familiar with projections, here&rsquo;s a brief explanation. Projection is an <em>invocable</em> entity which is applied to a range element before the algorithm&rsquo;s logic will use it. It can be a lambda, pointer-to-member (either data or function) or just a function pointer. Along this article I will use these two structures for examples:</p>
<pre><code>struct Y
{
    int a;
    int b;
    auto operator&lt;=&gt;(const Y&amp;) const = default;
};

struct X
{
    int x;
    Y y;
    auto operator&lt;=&gt;(const X&amp;) const = default;
};
</code></pre>
<p>If you don&rsquo;t know what <code>operator&lt;=&gt;</code> is, don&rsquo;t worry, in the context of this article you only need to know that it provides all the comparison operations ( <code>==, !=, &lt;, &lt;=, &gt;, &gt;=</code>) for both <code>X</code> and <code>Y</code>, they operate in a member-wise fashion. Ok, back to the subject, imagine that we want to sort a vector of <code>X</code> based on <code>X::x</code>. Here&rsquo;s how this can be done with the pre-ranges STL:</p>
<pre><code>std::vector&lt;X&gt; v;

std::sort(std::begin(v), std::end(v), [](const auto&amp; lhs, const auto&amp; rhs){
    return lhs.x &lt; rhs.x;
});
</code></pre>
<p>And here&rsquo;s how it can be done using projection and range-based algorithm:</p>
<pre><code>std::ranges::sort(v, std::less{}, &amp;X::x);
</code></pre>
<p>Now, <code>std::less</code> operates on <code>X::x</code> values but it&rsquo;s important to understand that the algorithm itself sorts original <code>X</code> elements, not just their <code>X::x</code> parts. Roughly:</p>
<pre><code>auto sort(auto range, auto compare, auto projection){
    // `it1` and `it2` are iterators from `range`
    // comparator is invoked on projected values
    if(compare(std::invoke(projection, *it1), std::invoke(projection, *it2))){
        // but moving/swapping is done on non-projected values
        std::ranges::iter_swap(it1, it2);
    }
    // ...
}
</code></pre>
<p>Projection provides clear separation of comparison logic from the element manipulation. These things are really orthogonal, it&rsquo;s nice that now we can keep them separate. And while the idea behind projection is great, its implementation has unpleasant side-effects which sometimes make developer lives harder.</p>
<hr>
<h2>Problems with existing design</h2>
<p>Several months ago I became involved in the <a href="https://wg21.link/p1708r5">P1708 Simple Statistical Functions</a> proposal. I needed those functions for my hobby project and started implementing them. This was my first experience in writing range-based API and that&rsquo;s how I got most of my unpleasant experience working with current projections design.</p>
<h3>Projections uglify function signatures</h3>
<p>In range-based API you usually have at least one range for which you need to support and hence provide a projection. For example, simplified signature of <code>copy_if</code> with removed return type and <code>O</code> type requirements:</p>
<pre><code>template&lt;ranges::input_range R, typename O,
    class Proj = std::identity,
    std::indirect_unary_predicate&lt;
        std::projected&lt;ranges::iterator_t&lt;R&gt;, Proj&gt;&gt; Pred&gt;
constexpr auto copy_if(R&amp;&amp; r, O result, Pred pred, Proj proj = {});
</code></pre>
<p>All range-based algorithms must have this additional function and template parameter that defaults to a no-op <code>std::identity</code> projection. Looks innocent? In P1708 we have weighted statistics so we use two ranges: one for values and one for weights, thus, we need one projection per range:</p>
<pre><code>template&lt;
    typename Values,
    typename ValuesProj = std::identity,
    typename Weights,
    typename WeightsProj = std::identity&gt;
constexpr auto mean(
    Values&amp;&amp; v, Weights&amp;&amp; w, ValuesProj proj1 = {}, WeightsProj proj2 = {});
</code></pre>
<p>Add to it more algorithm specific parameters like comparators and you&rsquo;ll get something like <code>std::ranges::merge()</code>:</p>
<pre><code>template&lt;
    ranges::input_range R1,
    ranges::input_range R2,
    std::weakly_incrementable O,
    class Comp = ranges::less,
    class Proj1 = std::identity,
    class Proj2 = std::identity&gt;
constexpr auto merge(R1&amp;&amp; r1, R2&amp;&amp; r2, O result,
        Comp comp = {}, Proj1 proj1 = {}, Proj2 proj2 = {});
</code></pre>
<p>I believe that in a good API default function arguments should be rare and their number should be small. Here, we have 6 parameters and 3 of them have default arguments. This signature is not good at all, we also will discuss usability of such API in following sections.</p>
<p>Another issue, though not so critical, is access to projected value type in order, for example, to constrain it. Recall the constraint from the <code>copy_if()</code>: <code>std::indirect_unary_predicate&lt;std::projected&lt;ranges::iterator_t&lt;R&gt;, Proj&gt;&gt; Pred</code>. It ensures that predicate <code>Pred</code> can be called with the result of applying projection <code>Proj</code> to the value of the iterator of a range <code>R</code>. It&rsquo;s understandable but still quite complex. In P1708R5 functions are supposed to work only on standard arithmetic types, a way to achieve it:</p>
<pre><code>template&lt;typename Range, typename Proj = std::identity&gt;
requires std::is_arithmetic_v&lt;
    std::remove_cvref_t&lt;
        std::indirect_result_t&lt;Proj&amp;, std::ranges::iter_value_t&lt;Range&gt;&gt;&gt;&gt;
double mean(Range&amp;&amp; r, Proj = {});
</code></pre>
<p>I mean, OK, it works and with some effort you can do it properly. But I don&rsquo;t like its complexity. Writing your own algorithms in the classic STL style was simple, writing them for ranges is not if you want to support projections properly.</p>
<hr>
<h3>Projections are not easily composable</h3>
<p>Imagine that you&rsquo;re implementing a range-based algorithm and you need to call another algorithm but with one more additional projection. For example, <a href="https://en.wikipedia.org/wiki/Geometric_mean#Relationship_with_logarithms">geometric mean</a> is usually implemented in terms of arithmetic mean of logarithms and final <code>std::exp()</code> of it. This requires combination of two projections, original one and <code>std::log()</code>:</p>
<pre><code>template&lt;typename R, typename P = std::identity&gt;
constexpr double geometric_mean(R&amp;&amp; r, P proj = {})
{
    const auto logs_mean = mean(
        std::forward&lt;R&gt;(r),
        [&amp;](const auto&amp; value)
        {
            return std::log(std::invoke(proj, value));
        });

    return std::exp(logs_mean);
}
</code></pre>
<p>It would be nice to move <code>std::log()</code> part to a separate independent projection but such a projection wouldn&rsquo;t be really independent because it needs to know about the preceding one:</p>
<pre><code>// to make it a reusable function-like object we again need this additional
// parameter everywhere
template&lt;typename P = std::identity&gt;
class log_proj{
public:
    explicit log_proj(P proj = {}): p{std::move(proj)}{}

    auto operator()(const auto&amp; value){
        return std::log(std::invoke(p, value));
    }
private:
    P p;
};
// with the above it's possible to write:
// const auto logs_mean = mean(r, log_proj{proj});

// and this is what I want as a client:
struct nice_log_proj{
    auto operator()(const auto&amp; value){
        return std::log(value);
    }
};
</code></pre>
<p>Of course, it&rsquo;s possible to create another utility to chain projections together like <code>mean(r, chain(std::move(proj), nice_log_proj{}))</code> but at the moment there&rsquo;s no standard tool for that. This problem also occurs when you want to sort <code>std::vector&lt;X&gt;</code> by <code>Y::a</code> member of <code>X::y</code>. In C++ it&rsquo;s not possible to get a pointer to a member of a member, <code>&amp;X::y::a</code> doesn&rsquo;t work, something like <code>chain(&amp;X::y, &amp;Y::a)</code> is needed.</p>
<hr>
<h3>Projections complicate caller&rsquo;s code</h3>
<p>Imagine a function with several default arguments:</p>
<pre><code>template&lt;typename R, typename P = std::identity&gt;
void f(R&amp;&amp; range, int x = 1, int y = 2, P p = {});
</code></pre>
<p>Because the projection is usually placed at the end of signature, if you need to use it, you have to specify all the default arguments by hand:</p>
<pre><code>f(v);               // without projection
f(v, 1, 2, &amp;X::x);  // with projection
</code></pre>
<p>What if the author of <code>f()</code> decides to change default arguments? Clients will be forced to rewrite the code to preserve &ldquo;default&rdquo; behavior. It&rsquo;s less painful with something like <code>std::ranges::sort()</code>:</p>
<pre><code>std::ranges::sort(v);   // without projection
std::ranges::sort(v, std::less{}, &amp;X::x);   // with projection
std::ranges::sort(v, {}, &amp;X::x);    // less verbose but less readable too
</code></pre>
<p>But now it&rsquo;s either too verbose with explicit <code>std::less{}</code> or less readable with <code>{}</code>. So the client is either forced to explicitly write arguments by hand or use less readable constructions if that&rsquo;s possible at all. Going back to weighted stats:</p>
<pre><code>template&lt;
    typename Values,
    typename Weights,
    typename ValuesProj = std::identity,
    typename WeightsProj = std::identity&gt;
constexpr auto mean(
    Values&amp;&amp; v, Weights&amp;&amp; w, ValuesProj proj1 = {}, WeightsProj proj2 = {});

mean(values, weights);    // no projections, great
mean(values, weights, &amp;X::x); // only value projection, OK
mean(values, weights, {}, &amp;X::x); // only weight projection, ugly :(
</code></pre>
<hr>
<h3>Root cause of all the problems</h3>
<p>From the interface point of view it&rsquo;s pretty simple, the problem is that range and projection represent a logically single entity but are passed to functions separately via distinct parameters. It&rsquo;s the same as for error-prone <code>f(const char* str, std::size_t len);</code> and we all know it&rsquo;s a bad way of doing things. Clients are forced to separate things, developers are forced to combine them back together, I want something better.</p>
<hr>
<h2>Projected ranges to the rescue</h2>
<p>I had quite a simple idea: range and projection should be combined into a single thing using some kind of view, e.g., <code>views::projection</code>. This would make all those projection-related parameters redundant, algorithms wouldn&rsquo;t care about them at all, they would only operate on a range itself, just like in classic STL. Here&rsquo;s what I wanted:</p>
<pre><code>// no projection-related parameters
auto sort(auto&amp;&amp; range, auto cmp = std::ranges::less{});

// sorts elements by `X::x` member, analog of current sort(v, {}, &amp;X::x);
sort(v | projection(&amp;X::x));

const auto log_proj = [](const auto value)
{
    return std::log(value);
};

// nested projections, actually, it's a nested range now
constexpr double geometric_mean(auto&amp;&amp; r)
{
    const auto logs_mean = mean(r | projection(log_proj{}));
    return std::exp(logs_mean);
}
</code></pre>
<p>Isn&rsquo;t it great? No more projection-related parameters, signatures are clean, everything is perfectly composable. It simplifies projections in the same way as ranges simplified usage and composition of iterator-based algorithms.</p>
<p>I call it a <em>projected range</em> because it combines range and projection. Such a range has very important property: its <code>operator*()</code> returns projected value, while copy/move/swap/assign operations should be performed on the whole underlying object. Immediately, another type of projection came to my mind, the so-called <code>narrow_projection</code>. All of its operations are performed on the projected part only. It&rsquo;s <em>narrow</em> in a sense that it represents only a narrow part of the object while <em>wide</em> <code>projection</code> represents a wider object behind it:</p>
<pre><code>std::vector&lt;X&gt; v{{3, {30, 300}}, {2, {20, 200}}, {1, {10, 100}}};

// sorts the whole X objects using &amp;X::x member
std::sort(v | projection(&amp;X::x));
// {{1, {10, 100}}, {2, {20, 200}}, {3, {30, 300}}}

// sorts only X::x
std::sort(v | narrow_projection(&amp;X::x), std::ranges::greater{});
// {{3, {10, 100}}, {2, {20, 200}}, {1, {30, 300}}}

// sorts X::y by Y::a
std::sort(
    v | narrow_projection(&amp;X::y) | projection(&amp;Y::a), std::ranges::greater{});
// {{3, {30, 300}}, {2, {20, 200}}, {1, {10, 100}}}
</code></pre>
<p>Delighted, I started to think how to implement it and it turned out to be a bit harder than I expected.</p>
<hr>
<h2>Implementation story</h2>
<p>If you don&rsquo;t know how range views work, here&rsquo;s the basic idea: all work is done inside custom &ldquo;smart&rdquo; iterators. For example, iterator for the most relevant to projections <code>views::transform</code> has <code>operator*()</code> which looks like this:</p>
<pre><code>class transform_view_iterator{
private:
    Iterator it;    // underlying iterator
    F f;            // transform function

public:
    decltype(auto) operator*(){
        return std::invoke(f, *it);
    }
    // ...
};
</code></pre>
<p>Other operations mostly take care of proper <code>it</code> moving. To implement <code>views::projection</code> we will need to implement a custom iterator, thus, we need first to understand how iterators work in C++20.</p>
<hr>
<h3>C++20 iterators overview</h3>
<p>Here&rsquo;s the brief overview of iterator-related types and operations. It&rsquo;s heavily based on articles/papers by Eric Niebler (<a href="http://ericniebler.com/2015/01/28/to-be-or-not-to-be-an-iterator/">0</a>, <a href="https://ericniebler.com/2015/02/03/iterators-plus-plus-part-1/">1</a>, <a href="https://ericniebler.com/2015/02/13/iterators-plus-plus-part-2/">2</a>, <a href="https://ericniebler.com/2015/03/03/iterators-plus-plus-part-3/">3</a>, <a href="https://ericniebler.github.io/std/wg21/D0022.html">4</a>). Read them if you want more details and reasoning behind current design.</p>
<p><code>iter_value_t/value_type</code> - the type of a value which the iterator represents. The value of this type can be copied/moved from the iterator.</p>
<p><code>iter_reference_t operator*()</code> - dereference operator, usually returns lvalue reference to <code>value_type</code> (but not required), must be convertible to <code>iter_value_t</code>.</p>
<p><code>iter_rvalue_reference_t iter_move(it)</code> - customization point for moving value out of iterator, usually returns rvalue reference to <code>value_type</code> (but not required). Also must be convertible to <code>iter_value_t</code>. If not defined by iterator, <code>std::move(*it)</code> is used.</p>
<p><code>void iter_swap(it1, it2)</code> - customization point for swapping values between two iterators. If not defined, performs <code>std::ranges::swap(*it1, *it2)</code> if possible, otherwise uses <code>iter_move()</code> to swap elements &ldquo;by-hand&rdquo;.</p>
<p><code>common_reference</code> requirements for readable iterators. Now comes the tricky part. As you might have noticed, <code>iter_value_t</code>, <code>iter_reference_t</code>, <code>iter_rvalue_reference_t</code> are not required to be as simple as <code>int</code>, <code>int&amp;</code> and <code>int&amp;&amp;</code> correspondingly. But there must be pairwise <code>common_reference</code>s to represent relationships between them. Basically, <code>common_reference&lt;T,U&gt;</code> is a type to which both <code>T</code> and <code>U</code> can be converted or bound, it&rsquo;s not required to be a true reference type.</p>
<pre><code>static_assert(std::same_as&lt;std::common_reference_t&lt;int&amp;, const int&amp;&gt;, const int&amp;&gt;);
static_assert(std::same_as&lt;std::common_reference_t&lt;int&amp;&amp;, int&amp;&amp;&gt;, int&amp;&amp;&gt;);
static_assert(std::same_as&lt;std::common_reference_t&lt;int&amp;&amp;, int&amp;&gt;, const int&amp;&gt;);
static_assert(std::same_as&lt;std::common_reference_t&lt;int&amp;, int&gt;, int&gt;);
</code></pre>
<p>You can find these requirements in the <a href="https://en.cppreference.com/w/cpp/iterator/indirectly_readable"><code>std::indirectly_readable</code></a> concept:</p>
<pre><code>template&lt;class In&gt;
concept __IndirectlyReadableImpl = // exposition only
requires(const In in) {
    typename std::iter_value_t&lt;In&gt;;
    typename std::iter_reference_t&lt;In&gt;;
    typename std::iter_rvalue_reference_t&lt;In&gt;;
    { *in } -&gt; std::same_as&lt;std::iter_reference_t&lt;In&gt;&gt;;
    { ranges::iter_move(in) } -&gt; std::same_as&lt;std::iter_rvalue_reference_t&lt;In&gt;&gt;;
} &amp;&amp;
std::common_reference_with&lt;
    std::iter_reference_t&lt;In&gt;&amp;&amp;, std::iter_value_t&lt;In&gt;&amp;&gt; &amp;&amp;
std::common_reference_with&lt;
    std::iter_reference_t&lt;In&gt;&amp;&amp;, std::iter_rvalue_reference_t&lt;In&gt;&amp;&amp;&gt; &amp;&amp;
std::common_reference_with&lt;
    std::iter_rvalue_reference_t&lt;In&gt;&amp;&amp;, const std::iter_value_t&lt;In&gt;&amp;&gt;;
</code></pre>
<p>Interestingly, there&rsquo;s no requirement that all these <code>common_reference</code>s must be the same type. In fact, they are not even required to be used and hence defined but they must be declared. Eric shows one example when <code>common_reference</code> might be useful, <code>unique_copy()</code> comparator parameter types. <code>unique_copy()</code> needs to copy <code>value_type</code> and then call comparator with this copy and the result of <code>operator*()</code> which is <code>iter_reference_t</code>. But the order of arguments is not specified. If for whatever reason your comparator cannot have templated parameters, you need to use <code>common_reference</code> for parameter types:</p>
<pre><code>auto unique_copy(Iterator first, Iterator last, auto d_first, auto comparator){
    // somewhere inside `unique_copy()`
    Iterator it = first;
    std::iter_value_t&lt;It&gt; copy = *it;   // copy current element
    ++it;
    comparator(copy, *it);  // compare it to the next one, it can be one way
    comparator(*it, copy);  // or the other
}

// client's code
auto generic_comparator = [](auto&amp; lhs, auto&amp; rhs){};   // no problems

// but if you need specific types, use common_reference
template&lt;std::indirectly_readable T&gt;
using iter_common_reference_t = std::common_reference_t&lt;
    std::iter_reference_t&lt;T&gt;,std::iter_value_t&lt;T&gt;&amp;&gt;;

auto non_generic_comparator = [](
    iter_common_reference_t&lt;T&gt; lhs, iter_common_reference_t&lt;T&gt; rhs){};
</code></pre>
<p>Note that since C++20, <code>iter_reference_t</code> is not required to be a true reference for any kind of iterator which effectively allows random-access proxy iterators.</p>
<p>Range-based versions of existing algorithms must be changed like this (current <code>libstdc++</code> still doesn&rsquo;t use <code>iter_move()</code>/<code>iter_swap()</code> in its range algorithms):</p>
<pre><code>Iterator it1, it2;

// pre C++20 algorithms:
using value_type = std::iterator_traits&lt;Iterator&gt;;
value_type copied = *it1;             // copy
value_type moved = std::move(*it1);   // move
std::iter_swap(it1, it2);             // swap

// C++20 algorithms:
using value_type = std::iter_value_t&lt;Iterator&gt;;
value_type copied = *it1;                       // copy
value_type moved = std::ranges::iter_move(it);  // move
std::ranges::iter_swap(it1, it2);               // swap
</code></pre>
<p>The main purpose of this design (as I understand it) is to allow proxy iterators of any kind, which, in theory, allows more &ldquo;indirect&rdquo; iterators and their usage with standard algorithms. Imaginary proxy-iterator must implement:</p>
<ul>
<li>corresponding to its category functions (<code>operator++()</code>, <code>operator[]</code>, etc.)</li>
<li>custom proxy-reference type which must have read/write/conversions to/from <code>value_type</code>, itself and <code>iter_rvalue_reference_t</code>
</li>
<li>custom <code>iter_move()</code> and <code>iter_swap()</code>
</li>
<li>specialize necessary <code>basic_common_reference</code> (a helper for <code>common_reference</code> described above) between its <code>value_type</code>, <code>iter_reference_t</code>, <code>iter_rvalue_reference_t</code> to a type which is at least declared</li>
</ul>
<hr>
<h3>Need for a better design</h3>
<p>Now, when you have a basic idea of what iterators can do in C++20, we can start to think about how <code>views::projection</code> should work. Recall usage example:</p>
<pre><code>sort(v | projection(&amp;X::x));    // sorts `v` by `X::x` member
</code></pre>
<p>For this to work, <code>operator*()</code> must return a reference-like thing which points to the projected value (<code>X::x</code> member in this case) so that the comparator will use it instead of the whole object. On the other hand, copy/move/swap/assign operations must operate on the non-projected object (<code>X</code>). In other words, we have two distinct types:</p>
<ul>
<li>
<code>value_type</code> - the type exposed through <code>operator*()</code>, projected type</li>
<li>
<code>iter_root_t</code> - the type of underlying object, root type</li>
</ul>
<p>and there&rsquo;s no logical relationship between them, i.e., there&rsquo;s no connection between <code>int x;</code> and <code>struct X;</code> types. One can argue that in fact we have <code>X</code> and <code>&amp;X::x</code> types and there is a <code>member-of</code> relation but in reality projection can also be a pure transformation, e.g., from <code>std::string</code> to <code>int</code> so any kind of relationship doesn&rsquo;t make sense here.<br>
In contrast, the existing design doesn&rsquo;t leave space for a second type (<code>iter_root_t</code>). It allows proxy-reference as <code>iter_reference_t</code> but it enforces strict relationships between it and <code>value_type</code> in terms of <code>common_reference</code> requirements. At most, it allows representing logically single <code>value_type</code> with two different types, like an advanced form of pointer. That&rsquo;s why related concepts are named like <code>indirectly_readable/writable/etc</code>, it&rsquo;s all about indirection mechanics, not true abstraction from one type to another.<br>
And even this indirection mechanism is over-complicated, I&rsquo;d say it&rsquo;s expert- <del>friendly</del> only utility. I mean, when Eric Niebler <a href="https://twitter.com/ericniebler/status/1363892564831711235">says it&rsquo;s hard</a>( you can check his implementation <a href="https://github.com/ericniebler/range-v3/blob/master/include/range/v3/iterator/basic_iterator.hpp">here</a>), how can you expect people to write their own iterators using it? It&rsquo;s hard because if you need to use proxy-reference, you need first to check and understand algorithm requirements on operations/conversions proxy-reference (<code>iter_reference_t</code>), <code>value_type</code> and <code>iter_rvalue_reference_t</code> should support and only then <em>try</em> to implement it.</p>
<p>To summarize, there are two main problems: over-complicated design and its inability to support true abstraction between two unrelated types. Now, let&rsquo;s fix it.</p>
<hr>
<h3>The next iteration of iterators</h3>
<p>In <a href="https://oleksandrkvl.github.io/2021/10/11/projected-ranges.html#projected-ranges">Projected ranges to the rescue</a> section I said that algorithms must operate on the <code>iter_root_t</code> values only, <code>value_type</code> should be used only when it&rsquo;s passed to customizable logic like comparators. Thus, we need to separate <code>value_type</code> API from <code>iter_root_t</code> API. Let&rsquo;s summarize what we have so far:</p>
<ul>
<li>
<code>operator*()</code> to get an lvalue reference or copy of <code>value_type</code>
</li>
<li>
<code>operator*()</code> to write <code>value_type</code>
</li>
<li>
<code>iter_move(it)</code> to get rvalue reference to <code>value_type</code>
</li>
<li>
<code>iter_swap(it1, it2)</code> to swap whatever we want, in our case it&rsquo;s <code>iter_root_t</code>
</li>
</ul>
<p>Now we need similar functions for <code>iter_root_t</code>:</p>
<ul>
<li>
<code>iter_copy_root(it)</code> to get an lvalue reference to <code>iter_root_t</code>, <code>iter_root_reference_t</code>
</li>
<li>
<code>iter_move_root(it)</code> to get an rvalue reference to <code>iter_root_t</code>, <code>iter_root_rvalue_reference_t</code>
</li>
</ul>
<p>And to simplify assignment:</p>
<ul>
<li>
<code>iter_assign_from(it, value)</code> to assign whatever is needed</li>
</ul>
<p>All these new functions are customization point objects (CPO) which means they are not required to be implemented if the iterator is happy with the default behavior. One of my goals was to preserve backward compatibility with all existing iterators so default implementations mostly forward to the old API. If you are not familiar with typical CPO implementation, the idea is quite simple: you call customized for a specific type function or the default implementation. The presence of a customized function is detected via ADL check (<code>has_adl_[cpo_name]</code> below). Implementation is located inside <code>struct</code> that&rsquo;s why in the code below <code>operator()(...)</code> is used instead of a plain function. <code>stdf</code> is a namespace name where I put all the new stuff, not a typo.</p>
<hr>
<h4>iter_copy_root()</h4>
<p>Returns lvalue reference to <code>iter_root_t</code>. Default behavior is to return the result of <code>operator*()</code>. I deliberately omit return type, <code>noexcept</code>-ness and constraint specifications since they are trivial, interested readers can find them in demo implementation.</p>
<pre><code>template&lt;typename From&gt;
constexpr decltype(auto) operator()(From&amp;&amp; from) const
{
    if constexpr(has_adl_iter_copy_root&lt;From&gt;)
    {
        return iter_copy_root(static_cast&lt;From&amp;&amp;&gt;(from));
    }
    else
    {
        return *from;
    }
}

// helper aliases
template&lt;typename T&gt;
using iter_root_t =
    std::remove_cvref_t&lt;decltype(stdf::iter_copy_root(std::declval&lt;T&gt;()))&gt;;

template&lt;typename T&gt;
using iter_root_reference_t = decltype(stdf::iter_copy_root(std::declval&lt;T&gt;()));

// usage example:
auto&amp; ref = stdf::iter_copy_root(it);
auto copy = stdf::iter_copy_root(it);
</code></pre>
<hr>
<h4>iter_move_root()</h4>
<p>Returns rvalue reference to underlying object. When not customized, can forward to <code>iter_move()</code> or to <code>iter_copy_root()</code>. Reason for this is simple: <code>iter_move_root()</code> is supposed to return rvalue reference to root type, if <code>iter_copy_root()</code> is not customized, it operates in terms of value type and <code>iter_move()</code> is responsible for moving it. This also preserves backward compatibility, for existing iterators <code>iter_copy_root()</code> is forwarded to <code>operator*()</code> and <code>iter_move_root()</code> to <code>iter_move()</code>.</p>
<pre><code>constexpr decltype(auto) operator()(From&amp;&amp; from) const
{
    if constexpr(has_adl_iter_move_root&lt;From&gt;)
    {
        return iter_move_root(static_cast&lt;From&amp;&amp;&gt;(from));
    }
    else if constexpr(
        iter_move_cpo::has_adl_iter_move&lt;From&gt; &amp;&amp;
        !iter_copy_root_cpo::has_adl_iter_copy_root&lt;From&gt;)
    {
        return stdf::iter_move(static_cast&lt;From&amp;&amp;&gt;(from));
    }
    else if constexpr(std::is_lvalue_reference_v&lt;
                            iter_root_reference_t&lt;From&gt;&gt;)
    {
        return std::move(
            stdf::iter_copy_root(static_cast&lt;From&amp;&amp;&gt;(from)));
    }
    else
    {
        return stdf::iter_copy_root(static_cast&lt;From&amp;&amp;&gt;(from));
    }
}

template&lt;typename T&gt;
using iter_root_rvalue_reference_t =
    decltype(stdf::iter_move_root(std::declval&lt;T&gt;()));

// usage example:
auto moved = stdf::iter_move_root(it);
</code></pre>
<hr>
<h4>iter_assign_from()</h4>
<p>It is responsible for assignment. Developer has full control over supported types. It&rsquo;s possible to introduce <code>iter_assign_value</code> and <code>iter_assign_root</code> but I don&rsquo;t know any use-case where it might be useful. Default behavior assigns to root:</p>
<pre><code>template&lt;typename To, typename From&gt;
constexpr void operator()(To&amp;&amp; to, From&amp;&amp; from) const
{
    if constexpr(has_adl_iter_assign_from&lt;To, From&gt;)
    {
        iter_assign_from(static_cast&lt;To&amp;&amp;&gt;(to), static_cast&lt;From&amp;&amp;&gt;(from));
    }
    else
    {
        stdf::iter_copy_root(static_cast&lt;To&amp;&amp;&gt;(to)) = static_cast&lt;From&amp;&amp;&gt;(from);
    }
}

// helper concept
template&lt;typename To, typename From&gt;
concept iter_assignable_from = requires(To&amp;&amp; to, From&amp;&amp; from)
{
    stdf::iter_assign_from(static_cast&lt;To&amp;&amp;&gt;(to), static_cast&lt;From&amp;&amp;&gt;(from));
};

// usage example:
stdf::iter_assign_from(it, T{});
</code></pre>
<hr>
<h4>iter_swap()</h4>
<p><code>iter_swap()</code> behaves almost like <code>std::ranges::iter_swap()</code> but it operates on root values, i.e., it uses <code>iter_copy_root()</code> instead of <code>operator*()</code> and <code>iter_move_root()</code>/<code>iter_assign_from()</code> instead of <code>iter_move()</code>/<code>operator=()</code>. I don&rsquo;t show implementation here because it&rsquo;s not so short, you can find it in the demo.</p>
<hr>
<h3>views::projection</h3>
<p>Now, when we have full control over the iterator&rsquo;s behavior, we can finally implement <code>views::projection</code> and <code>views::narrow_projection</code> and see how the new API simplifies custom iterator implementation. I will show only core parts of the iterator. We need to store current underlying iterator and a pointer to a parent view where projection function is stored:</p>
<pre><code>class Iterator
{
private:
    BaseIter current{};
    ParentView* parent{};
};
</code></pre>
<p>Core parts:</p>
<pre><code>constexpr decltype(auto) operator*() const
{
    return std::invoke(parent-&gt;fun, *current);
}

friend constexpr decltype(auto) iter_copy_root(const Iterator&amp; it)
{
    return stdf::iter_copy_root(it.current);
}

// enabled only if `BaseIter` has custom `iter_move_root`
friend constexpr decltype(auto) iter_move_root(const Iterator&amp; it)
{
    return stdf::iter_move_root(it.current);
}

// enabled only if `BaseIter` has custom `iter_swap`
friend constexpr void iter_swap(const Iterator&amp; x, const Iterator&amp; y)
{
    return stdf::iter_swap(x.current, y.current);
}

// enabled only if `BaseIter` has custom `iter_assign_from`
template&lt;typename T&gt;
friend constexpr void iter_assign_from(const Iterator&amp; it, T&amp;&amp; val)
{
    stdf::iter_assign_from(it.current, std::forward&lt;T&gt;(val));
}
</code></pre>
<p>As you can see, it&rsquo;s trivial, all of them are one-liners. <code>operator*()</code> returns the result of applying projection to the iterator&rsquo;s value. We don&rsquo;t need custom <code>iter_move()</code> because default implementation operates on the result of <code>operator*()</code>. Since we want copy/move/assign/swap operations to operate on the root value, we simply forward these calls to it. Note that the last three functions are enabled (using <code>requires</code>-clause) only in case when the underlying iterator customizes them. Otherwise, their default versions will operate on the basis of <code>iter_copy_root()</code> which is exactly what&rsquo;s needed. There&rsquo;s another reason why it&rsquo;s better to avoid customized versions of CPO-s when possible, it&rsquo;s described later in section <a href="https://oleksandrkvl.github.io/2021/10/11/projected-ranges.html#reducing-derefs">Reducing number of dereferences</a>.</p>
<hr>
<h3>views::narrow_projection</h3>
<p>It&rsquo;s even simpler, all we need is:</p>
<pre><code>constexpr decltype(auto) operator*() const
{
    return std::invoke(parent-&gt;fun, *current);
}
</code></pre>
<p>Because we don&rsquo;t want to expose the underlying root value to copy/move/swap/assign operations, everything else works by default.</p>
<hr>
<h3>Impact on algorithms</h3>
<p>Just like it was with C++20 iterator API, this one also requires algorithm authors to update implementations. Their requirements have to be updated to reflect usage of the new API. Changes to algorithms code are trivial, <code>operator*()</code> is still used for customizable logic like comparators but copy/move/swap/assign must be replaced with new functions. In the demo I implemented a couple of simple algorithms to test how the new design fits in and found no major problems.</p>
<pre><code>// read projected value
auto v1 = std::invoke(proj, *it);   // before
auto v2 = *it;                      // after

// copy underlying value, now copies `iter_root_t`
iter_value_t&lt;It&gt; copy1 = *it;               // before
iter_root_t&lt;It&gt; copy2 = iter_copy_root(it); // after

// move underlying value, now moves `iter_root_t`
iter_value_t&lt;It&gt; moved1 = iter_move(it);    // before
iter_root_t&lt;It&gt; moved = iter_move_root(it); // after

// assign to iterator
*it = val;                  // before
iter_assign_from(it, val);  // after
</code></pre>
<h4>Iterator-based versions of algorithms</h4>
<p>For some reason, all range-based algorithms also have iterator-based counterparts, e.g., <code>copy_if(Range r, Out o, Pred pred, Proj proj)</code> and <code>copy_if(I begin, S end, Out o, Pred pred, Proj proj)</code>. I don&rsquo;t know why they are needed at all when a pair of iterators can be converted into a range using <code>std::ranges::subrange</code> but they are here. Described <code>projection</code>/<code>narrow_projection</code> combine projection and a <em>range</em>. To remove projections from iterator-based signatures we need something like <code>projection_iterator</code>. It should work just like <code>projection_view::iterator</code> with addition of comparison functions with its root iterator to support cases like <code>std::ranges::sort(make_projection_iterator(std::begin(r), some_projection), std::end(r))</code>. Or this issue can be ignored at all, projections can be removed without introducing <code>projection_iterator</code>. It will force usage of <code>std::ranges::subrange</code> and projection on the resulting range.</p>
<hr>
<h3>Reducing number of dereferences</h3>
<p>While implementing algorithms, I found one interesting issue. Consider <code>copy_if()</code> algorithm. Here&rsquo;s <code>libstdc++</code> implementation:</p>
<pre><code>void copy_if(auto first, auto last, auto result, auto pred, auto proj)
{
    for (; first != last; ++first)
    {
        if (std::invoke(pred, std::invoke(proj, *first)))   // #1
        {
            *result = *first;   // #2
            ++result;
        }
    }
}
</code></pre>
<p>The subtle issue here is that <code>first</code> is dereferenced twice per iteration, first, to call the predicate, second, to copy its value to the output <code>result</code> iterator. As I told you before, <code>libstdc++</code> still uses old implementations for range-based algorithms and probably this version is OK for old-school iterators. But in a <code>ranges</code> world <code>operator*()</code> might do non-trivial things. For example, it might be a range which uses <code>views::transform</code> with <code>int -&gt; string</code> transformation. <code>range-v3</code> handles it better, whenever possible it stores and reuses the result of dereference:</p>
<pre><code>void copy_if(auto first, auto last, auto result, auto pred, auto proj)
{
    for (; first != last; ++first)
    {
        auto&amp;&amp; x = *first;     // dereference is done only once now
        if (std::invoke(pred, std::invoke(proj, x)))
        {
            *result = (decltype(x) &amp;&amp;)x;    // analog of std::forward&lt;...&gt;(x)
            ++result;
        }
    }
}
</code></pre>
<p>With that in mind, I wrote initial implementation using the new API:</p>
<pre><code>constexpr void copy_if(auto&amp;&amp; in, auto out, auto pred)
{
    auto first = std::ranges::begin(in);
    auto last = std::ranges::end(in);
    for(; first != last; ++first)
    {
        if(std::invoke(pred, *first))
        {
            iter_assign_from(out, iter_copy_root(first));
            ++out;
        }
    }
}
</code></pre>
<p>Explicit dereference is done only once here but recall that when <code>iter_copy_root()</code> is not customized by client, it falls back to <code>operator*()</code> so the above code transforms to:</p>
<pre><code>if(std::invoke(pred, *first))       // first dereference
{
    iter_assign_from(out, *first);  // second dereference
    ++out;
}
</code></pre>
<p>Taking into account that now <code>operator*()</code> might contain projection, I want to avoid calling it whenever possible. Also, standard algorithms guarantee a specific number of projection calls, any approach which cannot fulfill them would be useless. The fixed version would be:</p>
<pre><code>constexpr void copy_if(auto&amp;&amp; in, auto out, auto pred)
{
    auto first = std::ranges::begin(in);
    auto last = std::ranges::end(in);
    for(; first != last; ++first)
    {
        auto&amp;&amp; x = *first;
        if(std::invoke(pred, x))
        {
            if constexpr(has_adl_iter_copy_root&lt;decltype(first)&gt;)
            {
                // call customization point
                iter_assign_from(out, iter_copy_root(first));
            }
            else
            {
                // reuse `x`
                iter_assign_from(out, std::forward&lt;decltype(x)&gt;(x));
            }
            ++out;
        }
    }
}
</code></pre>
<p>Now when there&rsquo;s no customized <code>iter_copy_root()</code>, the dereferenced value can safely be reused. Obviously, having such an <code>if</code> statement in all algorithms for each call of <code>iter_copy_root</code>, <code>iter_move_root</code> and <code>iter_assign_from</code> would be too verbose. To simplify it, I added a second version for each CPO with additional <code>dereferenced</code> parameter at the end. Now <code>copy_if()</code> is shorter and dereferences only once:</p>
<pre><code>// second version of iter_copy_root()
template&lt;typename From&gt;
constexpr decltype(auto) operator()(
    From&amp;&amp; from, std::iter_reference_t&lt;From&gt;&amp; dereferenced) const
{
    if constexpr(has_adl_iter_copy_root&lt;From&gt;)
    {
        return iter_copy_root(static_cast&lt;From&amp;&amp;&gt;(from));
    }
    else if constexpr(std::is_lvalue_reference_v&lt;std::iter_reference_t&lt;From&gt;&gt;)
    {
        return dereferenced;
    }
    else
    {
        return std::move(dereferenced);
    }
}

constexpr void copy_if(auto&amp;&amp; in, auto out, auto pred)
{
    auto first = std::ranges::begin(in);
    auto last = std::ranges::end(in);
    for(; first != last; ++first)
    {
        auto&amp;&amp; x = *first;
        if(std::invoke(pred, x))
        {
            stdf::iter_assign_from(out, stdf::iter_copy_root(first, x));
            ++out;
        }
    }
}
</code></pre>
<p>As a side-effect, it reduces the number of dereferences for all currently existing iterators because they don&rsquo;t customize new CPOs. Usually, an optimizer is able to eliminate them but I like that now it&rsquo;s guaranteed by design with or without optimizations. The same problem exists for <code>iter_move()</code> and <code>iter_swap()</code> because when not customized, they dereference. At the end, I added a second version for them too. That&rsquo;s why in <code>views::projection</code> it&rsquo;s important to enable custom <code>iter_move_root()</code>, <code>iter_swap()</code>, <code>iter_assign_from()</code> only if they are customized by the underlying iterator. Customizing them unconditionally prevents reuse of dereferenced value.</p>
<hr>
<h3>root() method</h3>
<p>Sometimes we need to use the result of a generic algorithm with member function of a container. One such example is <code>remove_if()</code>. It returns a range of removed elements which are then <code>erase()</code>d using member function. The signature in <code>std::vector</code> is <code>constexpr iterator erase(const_iterator first, const_iterator last);</code>. The problem is that it takes <code>std::vector::const_iterator</code> and when we do:</p>
<pre><code>auto pv = v | projection(&amp;X::x);
auto removed = stdf::remove_if(pv, less_than&lt;int{3}&gt;{});
</code></pre>
<p><code>removed</code> contains a range of <code>projection_view::iterator</code> so we need a way to get the underlying iterator from it. It&rsquo;s possible to provide an implicit conversion for it but implicit conversions are always dangerous so for now I made it a normal member function. While the existing <code>base()</code> method of view iterators returns the last wrapped iterator, the new <code>root()</code> method returns the very first iterator in the projection chain. To make it generic, I added a <code>stdf::root(it)</code> free function which falls back to <code>it.root()</code> or just returns <code>it</code>. Now we can do:</p>
<pre><code>auto pv = v | projection(&amp;X::x);
auto removed = stdf::remove_if(pv, less_than&lt;int{3}&gt;{});
v.erase(stdf::root(removed.begin()), stdf::root(removed.end()));
</code></pre>
<hr>
<h3>Major flaw</h3>
<p>Unfortunately, after this article was published I received some feedback and realized that this design cannot replace projections when the algorithm operates on <em>input range/iterator</em>. The value represented by the input iterator is valid until the iterator is not incremented, all the copies of the iterator may be invalidated afterward. This restriction allows only single-pass algorithms. Consider one of the simplest algorithm, <code>max</code>:</p>
<pre><code>template&lt;typename R, typename C = std::less&lt;&gt;&gt;
auto max(R&amp;&amp; r, C pred = {}, auto proj)
{
    auto first = std::ranges::begin(r);
    auto last = std::ranges::end(r);
    
    std::ranges::range_value_t&lt;R&gt; result = *first;
    while(++first != last)
    {
        auto&amp;&amp; tmp = *first;
        if(invoke(pred, invoke(proj, result), invoke(proj, tmp))){
            result = (decltype(tmp) &amp;&amp;)tmp;
        }
    }

    return result;
}
</code></pre>
<p>Here, we need to store a copy of the current max element in the <code>result</code> variable. The projection <code>proj</code> is later applied to that copied element and that&rsquo;s the problem. In the proposed design, I assumed that projected value is always accessed through iterator, not through the root value. Here&rsquo;s the implementation using new design:</p>
<pre><code>template&lt;typename Rng, typename Cmp = std::less&lt;&gt;&gt;
stdf::iter_root_t&lt;std::ranges::iterator_t&lt;Rng&gt;&gt; max(Rng&amp;&amp; rng, Cmp pred = {})
{
    auto first = std::ranges::begin(rng);
    auto last = std::ranges::end(rng);
    using iterator_t = std::ranges::iterator_t&lt;Rng&gt;;

    std::iter_value_t&lt;iterator_t&gt; maxValue = *first;
    stdf::iter_root_t&lt;iterator_t&gt; root = stdf::iter_copy_root(first);
    while(++first != last)
    {
        auto&amp;&amp; tmp = *first;
        if(std::invoke(pred, result, tmp)){
            maxValue = (decltype(tmp) &amp;&amp;)tmp;
            root = stdf::iter_copy_root(first);
        }
    }
    return root;
}
</code></pre>
<p>As you can see, the only option is to copy both root and projected value which is not acceptable from the performance point of view. This problem exists only for input ranges and vanishes with forward ranges because for them it&rsquo;s safe to copy the iterator and call its <code>operator*()</code> to get the projected value. However, there are still plenty of algorithms which require only <code>input_range</code> so the proposed design cannot be used in its current form. Any potential projection replacement should be able to retrieve projected value from the root value, not from the iterator.</p>
<hr>
<h2>Other use-cases</h2>
<p>Introduced design significantly simplifies creation of non-trivial iterators. Because each aspect is handled separately, there&rsquo;s no need for tricky proxy reference objects. <code>common_reference</code> requirements are still there, now for both <code>value_type</code> and <code>iter_root_t</code>, but it&rsquo;s almost impossible to break them so clients shouldn&rsquo;t care or even know about their existence. For example, here&rsquo;s how infamous <code>std::vector&lt;bool&gt;::iterator</code> can be implemented:</p>
<pre><code>class Iterator
{
public:
    bool operator*();   // no need for proxy reference type
    // swaps bits
    friend void iter_swap(const Iterator&amp; lhs, const Iterator&amp; rhs);
    // assigns bit from bool value
    friend void iter_assign_from(const Iterator&amp; lhs, bool val);
};
</code></pre>
<p>Because there&rsquo;s no sense in true copy/move of a single bit, <code>iter_move()</code>, <code>iter_copy_root()</code>, <code>iter_move_root()</code> work in terms of <code>bool</code> value returned by <code>operator*()</code>. But <code>iter_swap()</code> needs to actually swap bit values and <code>iter_assign_from()</code> should assign <code>bool</code> to a specific bit, thus, they are customized. It&rsquo;s possible to achieve it with the existing design but it requires a custom proxy reference type and <code>basic_common_reference</code> specializations.</p>
<p>Another use-case might be various wrappers. Once I wanted to write a wrapper for <a href="https://rapidjson.org/">rapidjson</a> library. The main part was a wrapper class around <code>rapidjson::Value</code> and <code>rapidjson::Document::AllocatorType</code> which provided a more convenient interface similar to <a href="https://github.com/nlohmann/json">nlohmann/json</a>. Writing the wrapper itself was easy but I failed at the point when I needed to provide a random-access iterator which returns my wrapper by-value. In C++17 it was impossible to achieve simply because <code>operator*()</code> returns value instead of true reference and such iterator couldn&rsquo;t be a random-access one. In C++20 it should be possible, but again, requires a good understanding of proxy reference and <code>common_reference</code> requirements to implement it. With the proposed design it&rsquo;s straightforward: root type is <code>rapidjson::Value</code>, value type is a wrapper:</p>
<pre><code>class MyWrapper{
public:
    // interface methods...
private:
    rapidjson::Value* value;
    rapidjson::Document::AllocatorType* allocator;  // required for write ops
};

class MyIterator{
public:
    auto operator*(){
        return MyWrapper{*origIterator, allocator};
    }

    decltype(auto) iter_copy_root(){
        return *origIterator;
    }

    // other iterator methods...

private:
    rapidjson::Value::ValueIterator origIterator;
    rapidjson::Document::AllocatorType* allocator;
};
</code></pre>
<hr>
<h2>The role of std::views::transform</h2>
<p>Someone might think that <code>std::views::transform</code> can be used to combine projection and a range but currently it&rsquo;s mostly useless for that purpose. Its <code>iter_move()</code> operates on transformed value while <code>iter_swap</code> operates on the underlying non-transformed value so you can&rsquo;t use it with any algorithm that might use them both (like <code>sort()</code>, see the <a href="https://cplusplus.github.io/LWG/issue3520">issue 3520</a>). The proposed fix is to remove customized <code>iter_swap()</code> so that the default version will operate on transformed value. With that fix, <code>views::transform</code> will become almost the same as <code>narrow_projection</code> (the only difference is that, for unknown reason, <code>views::transform</code> has <a href="https://eel.is/c++draft/range.transform.iterator">customized</a> <code>iter_move()</code> which behaves exactly like <a href="https://eel.is/c++draft/iterator.cust.move#1.2">the default one</a>). But should it be used instead of <code>narrow_projection</code>? It&rsquo;s more like a naming question, I think that the name <code>transform</code> corresponds to a case when that&rsquo;s the real purpose of the code, just like classic <code>std::transform()</code>. The name <code>projection</code> better fits cases when you don&rsquo;t want to transform a range but only change its representation for an algorithm. Of course such a thing can still be called a <em>transformation</em>, it&rsquo;s hard to get a clear answer here.</p>
<hr>
<h2>Demo</h2>
<p>You can find the implementation of new CPO-s, <code>projection</code>, <code>narrow_projection</code> and a few test algorithms <a href="https://github.com/OleksandrKvl/projected_ranges/blob/master/src/main.cpp">here</a>. It&rsquo;s just a single file which you can <a href="https://godbolt.org/z/685Pcavb3">copy-paste to godbolt</a>, currently it works only with GCC-11 because Clang hasn&rsquo;t implemented ranges yet.</p>
<hr>
<h2>Wrap-up</h2>
<p>The main benefit of introduced design is the support of true abstraction behind the iterator. Projection is only one of its use cases and I believe that it can fully replace and enhance them, at the same time simplify creation of new iterators. Important point is that it&rsquo;s backward compatible, no need to change existing iterators, only the algorithms. Let me know what you think about it. Do you like this design? Would you like to see it in the standard? Can it solve some of your problems or will create new ones instead? Have I missed something else? Any meaningful feedback is welcome.</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Oleksandr Koval&rsquo;s blog<br>
<br>
January 13, 2022 at 05:35PM
